### ISSUE TITLE
Add full type signatures to instrinsics. Overloads? Generics? Tuple returns?

### ISSUE THREAD

**Issue by Bananattack** [wiz-lang/wiz#84]
Right now intrinsics do not perform any type-checking. They instead pass their arguments to the instruction pattern matcher which will accept or reject the pattern provided. There is no type-checking done, instead the arguments can be of any type that matches the lower-level operand patterns.

Do we add overload support? (allows many functions of different type signatures and arities, works for instruction intrinsics, regular functions, and inline functions, but doesn't play nice with the possibility of type inference/generic code later. "easier" to support, used in lots of languages, despite its drawbacks.)
Generic functions with type constraints? (more constrained, only one arity possible per function, and only works for inline functions and intrinsics, since Wiz would otherwise need a way of monomorphizing / instancing a generic. More complicated. But allows stuff like inference to be possible.)

Also, some functions like `divmod` need to return multiple values, but have to return `()` (aka `void`) because there is incomplete support for tuples. However, they take their arguments in the form `u8 in register`, so maybe it's clear that they can do anything with the designated store?

I don't know. dumping these thoughts here for now.

**Comment by IamRifki** [comment]
I think inference would be nice, maybe explicit overloading would be better?

**Comment by Bananattack** [comment]
Hmm, yeah, overloading is better for ease of implementation. Thought about this more. There still might be a way to add generics to the language later, without the classic overloads-and-generics resolution conflict that languages like C++ and C# have. eg. forbid a function an overloaded function and a generic function to exist of the same arity. Anyway, all problems that could try and solve whenever generics are thought about later.

Unlike C++, there no implicit integer type promotion, no weird int-to-pointer implicit casts, no implicits signed/unsigned promotions, etc. The only real exception is narrowing integer constants of type `iexpr` to a sized integer type when a target type is requested. So there aren't the same weird ambiguities, you must cast if something isn't an exact match. So overloading is always explicit. We could make it so integer literals must have a size suffix or casted whenever there are multiple overloads for different integer types. (Similar to how integer literals don't assume a size when used with implicitly typed constants)

If generics were added, the language could also make it a rule that if there are overloaded generics, there is no other function of the same arity. And use constraints to limit what the argument can do. So there might be a nice elegant way to add generics later.

Also to allow certain kinds of "macro" capabilities without having full-blown generics / template system, inline functions might be able to take arguments that are compile-time constants (ie. let expressions), and do compile-time computations with them at the inline site.

That's a lot of unrelated features, and I feel like my brain jumped around a lot and maybe this isn't the most coherent. So I'll try to summarize!

TL;DR: overloading could be easiest path for now, and doesn't exclude the possibility of doing generics later. (Generics might not be needed, either to do macro facilities)
